### ISSUE TITLE
Do not work on real rgbd data!!! 

### ISSUE THREAD

**Issue by longchao343** [zgojcic/3DSmoothNet#3]
I have tested the net on real rgbd datasets, the calculation progress was really slow, and could not registered which could be registed by fpfh.

**Comment by zgojcic** [comment]
Hello, 

could you please provide more information on which data you used with which parameters? 3DMatch data set is a "real" rgbd data set and the method works. If you can send us some point cloud that you used we could look into the problem and get back to you. With as little information as you provided we can sadly not help much. 

Best 

Zan

**Comment by longchao343** [comment]
Thanks for your quick reply. 
       I have used the 3dSmoothNet on matterport datasets which provides much more clean and accurate depth images than normal depth image. At first I randomly sample the keypoints uniformly on the cloud, and then extract the descriptors using 3dSmoothNet. But the registration result was really bad, I got a couple of identity transforms. I also tried to extract descriptors of every single point in the cloud, and then using the ransac to get the transformation like I do on the fpfh descriptors. But I aslo can not get good results.
       I also tried to use use the point clouds in the demo, but I found that only the provided keypoints could get good and accurate results. If I uniformly sampled the point cloud as keypoints or calculate all points' descriptors, the results would be very bad.
       Dose 3dSmoothNet need some kind of accurate keypoints? As far as I known, in point cloud registration, we can not always get very good and stable keypoints. Hope for your quick repply!
Best

**Comment by zgojcic** [comment]
Hello, 

thanks for your reply. It is hard say what is going wrong, could you please send me 2 point cloud fragments that you are trying to register to my email and I will have a look. 

3DSmoothNet does not require stable keypoints actually all the points in the 3DMatch and ETH data set are sampled randomly. The same goes for the points in the demo which should also work if you do new sampling.

Best,

Zan

**Comment by longchao343** [comment]
Hi,
&emsp;&emsp;I have send the raw tested point clouds to your email. The raw point clouds are very huge, about 60MB.

**Comment by longchao343** [comment]
@zgojcic Any feed backs?

**Comment by zgojcic** [comment]
@longchao343 We have performed some tests using 3DSmoothNet as well as other feature descriptors (FPFH, SHOT, 3DMatch). We are performing a thorough analysis and will present the results in the next days. As a first note it seems that due to symmetry the point-clouds you provided are quite a hard example, but 3DSmoothNet still outperforms all other descriptors.

More information to follow in the next days.

**Comment by longchao343** [comment]
Wow, awesome. Thanks very much! :)

**Comment by BlankCheng** [comment]
Hello. May I ask how to select the anchor points? I'm a bit confused with this part in paper because if randomly select the anchors in one cloud, how to find these points in the other cloud. And what if my point clouds have nothing overlapping. And how to compute overlapping rate if not annoted?
Many thanks
Best
Cheng

**Comment by BlankCheng** [comment]
Sorry to recomment, I want to make my confusion more clear. When training I have groudtruth transformation matrix and can select the nearest neighbour as positive sample and the BH loss and ensure their embeddings are similar. But when testing how can I guarantee my selecting interest points to be perfect enough. I may select one point on a tree top but I do not select this point on the other cloud on tree top.

**Comment by zgojcic** [comment]
Hello BlankCheng, 

so in the testing or inference part you cannot guarantee that the actual corresponding points are sampled. In fact in the evaluation on the 3DMatch and ETH data set we just select randomly 5000 point from one point cloud and 5000 from the other. For example if the point clouds only have 30% overlap a lot of those points actually lie in the area where there is no overlap and have no correspondence. Furthermore, also the points that do lie in the overlap region do not necessarily have a correspondence due to the random sampling.

This makes the task harder but it corresponds to the actual use cases where one cannot not say where the overlap region is. All local feature descriptors just rely on redundancy here (if not using keypoint extractors), i.e. if we sample enough points there should be some that have correspondence and hopefully we can recover them. 

You can check the supplementary material of the paper for more information on how we estimate the overlap of the point clouds.

Best 
Zan

**Comment by BlankCheng** [comment]
Thanks for your reply, it solves my problem! May I ask where to find the supplementary materials, it seems to not be in the repository.

So for example I extract 300 local descriptors from source cloud and target cloud respectively. Then use RANSAC every iteration randomly choosing 3 points or more as inliers from source cloud, and find correspondence point pairs by selecting the nearest neighbour on local descriptor in L2 distance. Finally use a SVD to solve transformation matrix based on correspondence and itrate many times util the result is good enough.

Is my description right? If it's right, will it be very time consuming to choose all possible inliers from 300 local descriptors isince it's a billion order of magnitude.

**Comment by choward232** [comment]
Hi, I also have a question. How do you specify the keypoints that you look for? Is it random? The demo dataset contains a set of seemingly random 1000 keypoints within each of the two pointclouds. I could not find an answer in the paper either! Thanks!

**Comment by zgojcic** [comment]
The keypoints are always selected randomly. In the demo file there are 1000 of them but in all the evaluation in the paper we use 5000 randomly selected within each point cloud.

**Comment by choward232** [comment]
Thanks for the quick response! Must have missed that part in the paper, I'll re-evaluate.

**Comment by roxanasoto** [comment]
Hi, I also have a question. How is the files .tfrecord composed? The paper says that its inputs are features in SDV voxels, or are representations of the 3DMatch Dataset

**Comment by roxanasoto** [comment]
Thanks for the quick response! I also another question, How is the file "validationData.npz" composed? The paper describes the split between dataset train and dataset test, but I don't found information about the split of dataset validation that you used, I supposed that it always contains an anchor positive par, negative par, and a label?

**Comment by zgojcic** [comment]
Sorry for the late response, this issue somehow stayed under our radar. In the paper we actually just validate on the randomly chosen training examples that were not used for training yet as the training dataset is so big. However, you can just do a random split of the data and it should be ok

**Comment by roxanasoto** [comment]
Could you explain to us, how or what techniques did you apply to generate the validation data?

**Comment by roxanasoto** [comment]
For the validation data do you also estimate a unique LRF?

**Comment by zgojcic** [comment]
Sorry I missed the first post. Yes we do exactly the same preprocessing (LRF estimation and SDV grid extraction) for the validation data

**Comment by zgojcic** [comment]
Closing due to inactivity. If you have any other questions simply reopen the issue.

**Comment by surabhi96** [comment]
Hi @zgojcic @roxanasoto , how do you generate training data with all the given samples in a scene? Do you combine a sample with every other sample and find the keypoints in the overlapping region? Thanks!
